### PR TITLE
Do browserify transforms as prepublish

### PR DESCRIPTION
**Pull request by nolanlawson**
This changes a few things:

- browserify transforms are now done as a prepublish step using `browerify-transform-cli`
- `lib` moved to `src`, `lib` gitignored
- `"files"` added to avoid accidentally publishing anything we don't want published
- `bundle-collapser/plugin` added for even smaller `dist/` builds.
- bumps `immediate` to latest

The benefits of doing the transforms as a prepublish rather than browserify transforms are huge. Besides supporting non-Browserify bundlers like Webpack, it also vastly improves the install and build times.

To give you an idea, I compared the current version of lie against my [lie-pretransform](https://github.com/nolanlawson/lie-pretransform). When `npm install`ed, `lie` has 46 total dependencies (`ls node_modules | wc -l`), takes 8.9 seconds to install on a Gigabit connection, and puts 6.9MB in node_modules. `lie-pretransform`, on the other hand, has only one total dependency (`immediate`), takes 1.9 seconds, and puts 100kB in node_modules.

Furthermore, because the transforms don't have to be re-run on the client side over and over again during browserification, projects build faster.  I clocked `browserify node_modules/lie` at **0.874s** and `browserify node_modules/lie-pretransform` at **0.497s**, meaning it builds almost twice as fast.

Let me know if there's anything I should improve about these changes. I realize it's a big PR to dump all at once, but I really like `lie` and wanted to improve its DX.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,40 +18,46 @@
...
+    "ncp": "^2.0.0",
     "phantomjs": "^1.9.9",
     "promises-aplus-tests": "calvinmetcalf/promises-tests#phantom",
-    "uglify-js": "^2.4.13"
+    "uglify-js": "^2.4.13",
+    "unreachable-branch-transform": "^0.3.0"
   },
   "dependencies": {
-    "es3ify": "^0.2.2",
-    "immediate": "~3.0.5",
```
Comment (calvinmetcalf) [inline package.json]: this is intentionally out of date to avoid bringing in some new features that we don't use in lie

**Comment by calvinmetcalf** [comment]
my current workflow when publish a module is to do

- `npm verison major|minor|patch` which bumps the version and creates a signed tag
- `git push origin master v?.?.?` (filled in with the current tag) to push the change and the tag I just created up to github
- `npm publish`

from my understanding this will break that workflow but I'd be happy to change my workflow but I'd really prefer to keep signing releases

File: package.json (hunk 1 above)
Comment (nolanlawson) [inline package.json]: OK, my bad.

**Comment by nolanlawson** [comment]
This should still give you exactly the same workflow, since the build will run as a `prepublish`. OTOH the Travis build is failing somehow, will investigate.

**Comment by calvinmetcalf** [comment]
yeah but the git tag won't have everything

**Comment by calvinmetcalf** [comment]
i think it's an issue with browserify-transform-cli looks like there is a trailing \r on the first line

**Comment by nolanlawson** [comment]
Ugh Windows problems. Where's the \r? I'll fix it.

As for the git tag issue, it's the same as for `dist` right? So would you prefer just to have `lib` not be .gitignored?

**Comment by nolanlawson** [comment]
OK I fixed browserify-transform-cli and made it so that `lib` works the same as `dist`, hopefully that'll work.

**Comment by calvinmetcalf** [comment]
wait is lib/browser.js? it's failing jshint ?

**Comment by nolanlawson** [comment]
OK, fixed now

**Comment by calvinmetcalf** [comment]
ok merged, with a few changes, mostly that I need to manually run build, also I switched to copyfiles instead of ncp on the grounds that I wrote it and am vain.  Overall it took me a bit to figure out what you were doing but it looks good once I figured it out.  Take a look and I'll publish as 3.1 latter today

**Pull request merged**